extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts

```diff
@@ -663,4 +692,76 @@ export class SimulationModeEthereumClientService {
+		if (balances.length === 0) return []
+		const tokenInterface = new ethers.Interface(['function balanceOf(address account) view returns (uint256)'])
+		const transactions = balances.map((balanceRequest, index) => {
+			const balanceOfCallData = stringToUint8Array(tokenInterface.encodeFunctionData('balanceOf', [addressString(balanceRequest.owner)]))
+			return {
+				type: '1559' as const,
+				from: MOCK_ADDRESS + BigInt(index) + 1n,
+				to: balanceRequest.token,
+				value: 0n,
+				input: balanceOfCallData,
+				maxFeePerGas: 0n,
+				maxPriorityFeePerGas: 0n,
+				accessList: [],
+				gas: 42000n,
+				chainId: 0n,
+				nonce: 0n,
+			}
+		})
+		const transactionQueueSize = transactionQueue.length
+		const response = await ethereumClientService.multicall(transactionQueue.concat(transactions), blockNumber)
+		if (response.length !== transactions.length + transactionQueueSize) throw new Error('Multicall length mismatch')
+		return balances.map((balance, index) => ({
+			token: balance.token,
+			owner: balance.owner,
+			balance: response[transactionQueueSize + index].statusCode === 'success' ? bytesToUnsigned(response[transactionQueueSize + index].returnValue) : undefined
+		}))
+	}	
+
+	static readonly getAddressesInteractedWithERC20s = (events: MulticallResponseEventLogs): { token: bigint, owner: bigint }[] => {
+		const erc20ABI = [
+			'event Transfer(address indexed from, address indexed to, uint256 value)',
+			'event Approval(address indexed owner, address indexed spender, uint256 value)',
+		]
+		const erc20 = new ethers.Interface(erc20ABI)
+		const tokenOwners: { token: bigint, owner: bigint }[] = []
+		for (const log of events) {
+			const parsed = erc20.parseLog({ topics: log.topics.map((x) => bytes32String(x)), data: dataStringWith0xStart(log.data) })
+			if (parsed === null) continue
+			switch (parsed.name) {
+				case 'Approval':
+				case 'Transfer': {
+					tokenOwners.push({ token: log.loggersAddress, owner: EthereumAddress.parse(parsed.args[0]) })
+					tokenOwners.push({ token: log.loggersAddress, owner: EthereumAddress.parse(parsed.args[1]) })
+					break
+				}
+				default: throw new Error(`wrong name: ${ parsed.name }`)				
+			}
+		}
+		return tokenOwners
+	}
+
+	static readonly getTokenBalancesAfter = async (
+		ethereumClientService: EthereumClientService,
+		signedTxs: EthereumSignedTransaction[] = [],
+		multicallResult: MulticallResponse,
+		blockNumber: bigint,
+	) => {
+		const tokenBalancesAfter: TokenBalancesAfter[] = []
+		for (let resultIndex = 0; resultIndex < multicallResult.length; resultIndex++) {
+			const singleResult = multicallResult[resultIndex]
+			const balances = await SimulationModeEthereumClientService.getTokenBalances(
+				ethereumClientService,
+				signedTxs.slice(0, resultIndex),
+				SimulationModeEthereumClientService.getAddressesInteractedWithERC20s(singleResult.statusCode === 'success' ? singleResult.events : []),
+				blockNumber
+			)
```

Comment:
This code looks almost identical to the code above, would there be valuae in extracting out a helper function?
```suggestion
			const transactionStack = signedTxs.slice(0, resultIndex)
			const eventsInTopTransaction = SimulationModeEthereumClientService.getAddressesInteractedWithERC20s(singleResult.statusCode === 'success' ? singleResult.events : [])
			const balances = await SimulationModeEthereumClientService.getTokenBalances(ethereumClientService, transactionStack, eventsInTopTransaction, blockNumber)
```